crates/executor/src/state_reader.rs

```diff
@@ -21,51 +21,70 @@ use tracing::warn;
 
 use super::felt::{IntoFelt, IntoStarkFelt};
 
+type ContractCache = Arc<Mutex<SizedCache<StarknetClassHash, CachedContractClass>>>;
+
+struct CachedContractClass {
+    pub definition: ContractClass,
+    pub cached_on_height: Option<BlockNumber>,
+}
+
 /// A `StateReader` wrapper designed to cache the compiled contract classes.
 /// Contract classes are immutable once deployed so caching should not cause any side effect.
 pub(super) struct LruCachedReader<R>
 where
     R: StateReader,
 {
-    compiled_class_cache: GlobalContractCache,
+    compiled_class_cache: ContractCache,
     inner_reader: R,
+    height: Option<BlockNumber>,
 }
 
 impl<R> LruCachedReader<R>
 where
     R: StateReader,
 {
-    /// Build a `CachedState` on top of an `LruCachedReader`, sharing the same underlying cache.
-    pub fn _new_cached_state(inner_reader: R) -> anyhow::Result<CachedState<LruCachedReader<R>>> {
+    /// Build a `CachedState` on top of an `LruCachedReader`. The `CachedState`'s global cache is separate from the LRU cache.
+    pub fn new_cached_state(
+        inner_reader: R,
+        height: Option<BlockNumber>,
+    ) -> anyhow::Result<CachedState<LruCachedReader<R>>> {
+        lazy_static::lazy_static!(
+            static ref REGULAR_CACHE: GlobalContractCache = {
+                let inner = SizedCache::with_size(128);
+                GlobalContractCache(Arc::new(Mutex::new(inner)))
+            };
+        );
+
         // `CachedState` already has a [move_classes_to_global_cache](https://docs.rs/blockifier/0.3.0-rc0/blockifier/state/cached_state/struct.CachedState.html#method.move_classes_to_global_cache)
         // method that we might want to use instead, but relying on it would make this cache much
         // less self-contained as these calls would have to be made in various places.
         // One of the reasons for this is that we can't wrap the `CachedReader` in another layer of
         // the `State` trait as `blockifier` explicitly requires a `CachedReader` in the signature
         // of some methods we use.
-        let reader = Self::_new(inner_reader)?;
-        let cache = reader.compiled_class_cache.clone();
-        Ok(CachedState::new(reader, cache))
+        let reader = Self::new(inner_reader, height)?;
+        Ok(CachedState::new(reader, REGULAR_CACHE.clone()))
```

Comment:
This is probably wrong: with two caches, I'm quite sure a bug similar to the original may arise. Slightly different cause, same symptom.

Reply:
Sergey susggests another place for the fix:
https://github.com/eqlabs/pathfinder/blob/3d260893496d5758b23463b3f06547ca230ce5ec/crates/executor/src/state_reader.rs#L247

May be better & simpler.